File: sections/35_math_mixed_contest/problems/p_subarray_gcd_sum/solution.cpp
#include <bits/stdc++.h>
using namespace std;

const long long MOD = 1'000'000'007;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;
    vector<pair<int, long long>> previous;
    long long answer = 0;

    for (int i = 0; i < n; ++i) {
        int x;
        cin >> x;
        vector<pair<int, long long>> current{{x, 1}};
        for (auto [g, count] : previous) {
            int next_gcd = gcd(g, x);
            if (current.back().first == next_gcd)
                current.back().second += count;
            else
                current.push_back({next_gcd, count});
        }
        for (auto [g, count] : current)
            answer = (answer + 1LL * g * count) % MOD;
        previous = move(current);
    }

    cout << answer << '\n';
}

File: sections/35_math_mixed_contest/problems/p_subarray_gcd_sum/solution.py
import math
import sys

MOD = 1_000_000_007


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    n = data[0]
    previous = []
    answer = 0

    for x in data[1:1 + n]:
        current = [(x, 1)]
        for g, count in previous:
            next_gcd = math.gcd(g, x)
            if current[-1][0] == next_gcd:
                current[-1] = (next_gcd, current[-1][1] + count)
            else:
                current.append((next_gcd, count))
        answer = (answer + sum(g * count for g, count in current)) % MOD
        previous = current

    print(answer)


if __name__ == "__main__":
    main()
